[FEATURE] Explain listening to an event in how-to (#2326)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ExtensionArchitecture/HowTo/Events/Index.rst

.. include:: /Includes.rst.txt
.. index:: pair: Extension development; Events
.. _extension-development-events:

======
Events
======

:ref:`PSR-14 events <EventDispatcher>` can be used to extend the TYPO3 Core
or third-party extensions.

You can find a complete list of events provided by the TYPO3 Core in the
following chapter: :ref:`eventlist`.

> Events provided by third-party extensions should be described in the extensions

Reviewer: ```suggestion
Events provided by third-party extensions should be described in the extension's
```

> manual. You can also search for events by searching for classes that inject the

Reviewer: ```suggestion
manual. You can also search for events by looking for classes that inject the
```

This prevents using two time "search" in the sentence.

:ref:`EventDispatcherInterface`

.. _extension-development-event-listener:

Listen to an event
==================

> If you want to use an event provided by the Core or a third party extension,

Reviewer: ```suggestion
If you want to use an event provided by the Core or a third-party extension,
```

create a PHP class with a method `public __invoke(SomeCoolEvent $event)`
that accepts an object of the event class as

> argument. It is possible use another method name but you have to configure

Reviewer: ```suggestion
argument. It is possible to use another method name but you have to configure
```

the name in the :file:`Configuration/Services.yaml` or it is not found.

> It is best practice to use a descriptive class name that ends on

Reviewer: Hmm, the Core does not adhere to that best practise: https://github.com/TYPO3/typo3/tree/main/typo3/sysext/indexed_search/Classes/EventListener
It is already in a folder EventListener why should this the best practise as this might be produce long class names which may be hard to read, like in the given example then: `FrontendGenerationPageIndexingTriggerEventListener`. And it has no additional value.

Author: true

:file:`EventListener` and to put it in the namespace
:php:`MyVendor\MyExtension\EventListener`.

> .. literalinclude:: _Joh316PasswordInvalidEventListener.php

Reviewer: Maybe add a paragraph what is so special about "joh316" as password and why ist hosuld be prevented to use it (making some TYPO3 archaeology and history).

Author: added it as comment in the listener class

   :language: php

Then register the event in your extension's :file:`Configuration/Services.yaml`:

.. literalinclude:: _EventServices.yaml
   :language: yaml

> The file:`Configuration/Services.yaml` also allows to define a different

Reviewer: ```suggestion
Additionally, the :file:`Configuration/Services.yaml` file allows to define a different
```

> method name for the EventListener class and to influence the order in which

Reviewer: why "EventListener" and not just "event listener"?

> method name for the EventListener class and to influence the order in which

Author: ```suggestion
method name for the event listener class and to influence the order in which
```

events are loaded. See :ref:`EventDispatcherRegistration` for details.

Dispatch an event
=================

> You can dispatch events in your own extension`s code to enable other extensions

Reviewer: ```suggestion
You can dispatch events in your own extension's code to enable other extensions
```
